File: MD_sample/C_procedural/force_omp.c
#include <stdlib.h>
#include <math.h>
#include <omp.h>
#include "md.h"

double dnint(double x);

void force_omp(particle *q)
{
    unsigned i, j, k;
    double r2, dx[3], coef, x;
    double **tmp;

/* force initialization */
    for(i=0; i<Npt; i++) for (k=0;k<3;k++) q[i].ff[k] = 0.0;

/* coupled loop for many-body interaction */
#pragma omp parallel default(shared) private(tmp, i, k)
    {
	tmp = malloc(Npt*sizeof **tmp);
	for (i=0;i<Npt;i++) tmp[i] = malloc(3*sizeof *tmp);
	for (i=0;i<Npt;i++) for (k=0;k<3;k++) tmp[i][k] = 0.0;
#pragma omp for private(j,r2,x,dx,coef)
	for (i=0; i<Npt-1; i++) {
	    for (j=i+1; j<Npt; j++) {
		r2 = 0.0;
		for (k=0; k<3; k++) {
		    x = q[i].xx[k] - q[j].xx[k];
		    dx[k] = x - box * dnint(x/box);
		    r2 += dx[k]*dx[k];
		}
		if (r2 < rcut2) {
		    coef = 1./(r2*sqrt(r2));
		    for (k=0; k<3; k++) {
			tmp[i][k] += coef * dx[k];
			tmp[j][k] -= coef * dx[k];
		    }
		}
	    }
	}
#pragma omp critical
	{
	    for (i=0;i<Npt;i++) for (k=0;k<3;k++) q[i].ff[k] += tmp[i][k];
	}
	for (i=0;i<Npt;i++) free(tmp[i]);
	free(tmp);
    }
}
  

File: MD_sample/F90/force_omp.f90
!
! Parallel force-field routine
!##################################################
subroutine force_omp
  use md
  implicit none
  !
  integer(di):: i, j, ierr
  real   (dp):: r2, dx(3), coef
  real   (dp), allocatable:: tmp(:,:)
  !
  ! Initialization of force components
  do i=1_di, Npt
     q(i)%ff(:) = 0.0_dp
  end do
  !
  ! Force update
  !$omp parallel &
  !$omp default(shared) private(tmp, ierr)
  allocate(tmp(3_di,Npt), stat=ierr)
  tmp(:,:) = 0.0_dp
  !$omp do private(i,j, r2,dx,coef)&
  !$omp schedule(dynamic)
  do i=1_di, Npt-1_di
     do j=i+1_di, Npt
        !
        r2 = 0.0_dp
        dx(:) = q(i)%xx(:) - q(j)%xx(:)
        dx(:) = dx(:) - l_box*dnint(dx(:)/l_box)
        r2 = r2 + dx(1)*dx(1) + dx(2)*dx(2) + dx(3)*dx(3)
        coef = 1.0_dp/r2/dsqrt(r2)
        !
        tmp(:,i) = tmp(:,i) + coef*dx(:)
        tmp(:,j) = tmp(:,j) - coef*dx(:)
     end do
  end do
  !$omp end do
  !$omp critical
  do i=1_di, Npt
     q(i)%ff(:) = q(i)%ff(:) + tmp(:,i)
  end do
  !$omp end critical
  deallocate(tmp, stat=ierr)
  !$omp end parallel
  return
end subroutine force_omp
